File: src/main/java/seremis/geninfusion/soul/TraitHandler.java
package seremis.geninfusion.soul;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.IEntityLivingData;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.ChunkCoordinates;
import net.minecraft.util.DamageSource;
import seremis.geninfusion.api.soul.IEntitySoulCustom;
import seremis.geninfusion.api.soul.ITrait;
import seremis.geninfusion.api.soul.SoulHelper;

public class TraitHandler {

    public static void entityUpdate(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.onUpdate(entity);
        }
    }

    public static boolean interact(IEntitySoulCustom entity, EntityPlayer player) {
        boolean flag = false;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            if(trait.interact(entity, player)) {
                flag = true;
            }
        }
        return flag;
    }

    public static void entityDeath(IEntitySoulCustom entity, DamageSource source) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.onDeath(entity, source);
        }
    }

    public static void onKillEntity(IEntitySoulCustom entity, EntityLivingBase killed) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.onKillEntity(entity, killed);
        }
    }

    public static boolean attackEntityFrom(IEntitySoulCustom entity, DamageSource source, float damage) {
        boolean flag = false;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            if(trait.attackEntityFrom(entity, source, damage)) {
                flag = true;
            }
        }
        return flag;
    }

    public static void attackEntity(IEntitySoulCustom entity, Entity entityToAttack, float distance) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.attackEntity(entity, entityToAttack, distance);
        }
    }

    public static IEntityLivingData spawnEntityFromEgg(IEntitySoulCustom entity, IEntityLivingData data) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.onSpawnWithEgg(entity, data);
        }
        return data;
    }

    public static void playSoundAtEntity(IEntitySoulCustom entity, String name, float volume, float pitch) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.playSound(entity, name, volume, pitch);
        }
    }

    public static void damageEntity(IEntitySoulCustom entity, DamageSource source, float damage) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.damageEntity(entity, source, damage);
        }
    }

    public static void updateAITick(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.updateAITick(entity);
        }
    }

    public static void firstTick(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.firstTick(entity);
        }
    }

    public static boolean attackEntityAsMob(IEntitySoulCustom entity, Entity entityToAttack) {
        boolean flag = false;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            if(trait.attackEntityAsMob(entity, entityToAttack)) {
                flag = true;
            }
        }
        return flag;
    }

    public static Entity findPlayerToAttack(IEntitySoulCustom entity) {
        Entity flag = null;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            Entity tmp = trait.findPlayerToAttack(entity);
            if(tmp != null) {
                flag = tmp;
            }
        }
        return flag;
    }

    public static float applyArmorCalculations(IEntitySoulCustom entity, DamageSource source, float damage) {
        float flag = 0.0F;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            float tmp = trait.applyArmorCalculations(entity, source, damage);
            if(tmp != 0.0F) {
                flag = tmp;
            }
        }
        return flag;
    }

    public static float applyPotionDamageCalculations(IEntitySoulCustom entity, DamageSource source, float damage) {
        float flag = 0.0F;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            float tmp = trait.applyPotionDamageCalculations(entity, source, damage);
            if(tmp != 0.0F) {
                flag = tmp;
            }
        }
        return flag;
    }

    public static void damageArmor(IEntitySoulCustom entity, float damage) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.damageArmor(entity, damage);
        }
    }

    public static void setOnFireFromLava(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.setOnFireFromLava(entity);
        }
    }

    public static float getBlockPathWeight(IEntitySoulCustom entity, int x, int y, int z) {
        float flag = 0.0F;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            float tmp = trait.getBlockPathWeight(entity, x, y, z);
            if(tmp != 0.0F) {
                flag = tmp;
            }
        }
        return flag;
    }

    public static void updateEntityActionState(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.updateEntityActionState(entity);
        }
    }

    public static void updateWanderPath(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.updateWanderPath(entity);
        }
    }

    public static void attackEntityWithRangedAttack(IEntitySoulCustom entity, EntityLivingBase target, float distanceModified) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.attackEntityWithRangedAttack(entity, target, distanceModified);
        }
    }

    @SideOnly(Side.CLIENT)
    public static void render(IEntitySoulCustom entity, float timeModifier, float limbSwing, float specialRotation, float rotationYawHead, float rotationPitch, float scale) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.render(entity, timeModifier, limbSwing, specialRotation, rotationYawHead, rotationPitch, scale);
        }
    }

    public static boolean isWithinHomeDistanceCurrentPosition(IEntitySoulCustom entity) {
        boolean flag = false;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            if(trait.isWithinHomeDistanceCurrentPosition(entity)) {
                flag = true;
            }
        }
        return flag;
    }

    public static boolean isWithinHomeDistance(IEntitySoulCustom entity, int x, int y, int z) {
        boolean flag = false;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            if(trait.isWithinHomeDistance(entity, x, y, z)) {
                flag = true;
            }
        }
        return flag;
    }

    public static ChunkCoordinates getHomePosition(IEntitySoulCustom entity) {
        ChunkCoordinates flag = null;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            ChunkCoordinates coords = trait.getHomePosition(entity);
            if(coords != null) {
                flag = coords;
            }
        }
        return flag;
    }

    public static void setHomeArea(IEntitySoulCustom entity, int x, int y, int z, int maxDistance) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.setHomeArea(entity, x, y, z, maxDistance);
        }
    }

    public static float getMaxHomeDistance(IEntitySoulCustom entity) {
        float flag = 0.0F;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            float dist = trait.getMaxHomeDistance(entity);
            if(dist != 0.0F) {
                flag = dist;
            }
        }
        return flag;
    }

    public static void detachHome(IEntitySoulCustom entity) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.detachHome(entity);
        }
    }

    public static boolean hasHome(IEntitySoulCustom entity) {
        boolean flag = false;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            if(trait.hasHome(entity)) {
                flag = true;
            }
        }
        return flag;
    }

    public static void writeToNBT(IEntitySoulCustom entity, NBTTagCompound compound) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.writeToNBT(entity, compound);
        }
    }

    public static void readFromNBT(IEntitySoulCustom entity, NBTTagCompound compound) {
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            trait.readFromNBT(entity, compound);
        }
    }

    public static String getEntityTexture(IEntitySoulCustom entity) {
        String flag = null;
        for(ITrait trait : SoulHelper.traitRegistry.getTraits()) {
            String tex = trait.getEntityTexture(entity);
            if(tex != null && !tex.equals("")) {
                flag = tex;
            }
        }
        return flag;
    }
}

File: src/main/java/seremis/geninfusion/soul/TraitHandler.scala
package seremis.geninfusion.soul

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.effect.EntityLightningBolt
import net.minecraft.entity.{Entity, EntityLivingBase, IEntityLivingData}
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{ChunkCoordinates, DamageSource}
import seremis.geninfusion.api.soul.{IEntitySoulCustom, SoulHelper}
import seremis.geninfusion.soul.entity.EntitySoulCustomTrait

object TraitHandler {

    def entityUpdate(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.onUpdate(entity)
        }
    }

    def interact(entity: IEntitySoulCustom, player: EntityPlayer): Boolean = {
        var flag = false
        for (trt <- SoulHelper.traitRegistry.getTraits if trt.interact(entity, player)) {
            flag = true
        }
        flag
    }

    def entityDeath(entity: IEntitySoulCustom, source: DamageSource) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.onDeath(entity, source)
        }
    }

    def onKillEntity(entity: IEntitySoulCustom, killed: EntityLivingBase) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.onKillEntity(entity, killed)
        }
    }

    def attackEntityFrom(entity: IEntitySoulCustom, source: DamageSource, damage: Float): Boolean = {
        var flag = false
        for (trt <- SoulHelper.traitRegistry.getTraits if trt.attackEntityFrom(entity, source, damage)) {
            flag = true
        }
        flag
    }

    def attackEntity(entity: IEntitySoulCustom, entityToAttack: Entity, distance: Float) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.attackEntity(entity, entityToAttack, distance)
        }
    }

    def spawnEntityFromEgg(entity: IEntitySoulCustom, data: IEntityLivingData): IEntityLivingData = {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.onSpawnWithEgg(entity, data)
        }
        data
    }

    def playSoundAtEntity(entity: IEntitySoulCustom,
                          name: String,
                          volume: Float,
                          pitch: Float) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.playSound(entity, name, volume, pitch)
        }
    }

    def damageEntity(entity: IEntitySoulCustom, source: DamageSource, damage: Float) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.damageEntity(entity, source, damage)
        }
    }

    def updateAITick(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.updateAITick(entity)
        }
    }

    def firstTick(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.firstTick(entity)
        }
    }

    def attackEntityAsMob(entity: IEntitySoulCustom, entityToAttack: Entity): Boolean = {
        var flag = false
        for (trt <- SoulHelper.traitRegistry.getTraits if trt.attackEntityAsMob(entity, entityToAttack)) {
            flag = true
        }
        flag
    }

    def findPlayerToAttack(entity: IEntitySoulCustom): Entity = {
        var flag: Entity = null
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val tmp = trt.findPlayerToAttack(entity)
            if (tmp != null) {
                flag = tmp
            }
        }
        flag
    }

    def applyArmorCalculations(entity: IEntitySoulCustom, source: DamageSource, damage: Float): Float = {
        var flag = 0.0F
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val tmp = trt.applyArmorCalculations(entity, source, damage)
            if (tmp != 0.0F) {
                flag = tmp
            }
        }
        flag
    }

    def applyPotionDamageCalculations(entity: IEntitySoulCustom, source: DamageSource, damage: Float): Float = {
        var flag = 0.0F
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val tmp = trt.applyPotionDamageCalculations(entity, source, damage)
            if (tmp != 0.0F) {
                flag = tmp
            }
        }
        flag
    }

    def damageArmor(entity: IEntitySoulCustom, damage: Float) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.damageArmor(entity, damage)
        }
    }

    def setOnFireFromLava(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.setOnFireFromLava(entity)
        }
    }

    def getBlockPathWeight(entity: IEntitySoulCustom,
                           x: Int,
                           y: Int,
                           z: Int): Float = {
        var flag = 0.0F
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val tmp = trt.getBlockPathWeight(entity, x, y, z)
            if (tmp != 0.0F) {
                flag = tmp
            }
        }
        flag
    }

    def updateEntityActionState(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.updateEntityActionState(entity)
        }
    }

    def updateWanderPath(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.updateWanderPath(entity)
        }
    }

    def attackEntityWithRangedAttack(entity: IEntitySoulCustom, target: EntityLivingBase, distanceModified: Float) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.attackEntityWithRangedAttack(entity, target, distanceModified)
        }
    }

    @SideOnly(Side.CLIENT)
    def render(entity: IEntitySoulCustom,
               timeModifier: Float,
               limbSwing: Float,
               specialRotation: Float,
               rotationYawHead: Float,
               rotationPitch: Float,
               scale: Float) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.render(entity, timeModifier, limbSwing, specialRotation, rotationYawHead, rotationPitch,
                scale)
        }
    }

    def isWithinHomeDistanceCurrentPosition(entity: IEntitySoulCustom): Boolean = {
        var flag = false
        for (trt <- SoulHelper.traitRegistry.getTraits if trt.isWithinHomeDistanceCurrentPosition(entity)) {
            flag = true
        }
        flag
    }

    def isWithinHomeDistance(entity: IEntitySoulCustom,
                             x: Int,
                             y: Int,
                             z: Int): Boolean = {
        var flag = false
        for (trt <- SoulHelper.traitRegistry.getTraits if trt.isWithinHomeDistance(entity, x, y, z)) {
            flag = true
        }
        flag
    }

    def getHomePosition(entity: IEntitySoulCustom): ChunkCoordinates = {
        var flag: ChunkCoordinates = null
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val coords = trt.getHomePosition(entity)
            if (coords != null) {
                flag = coords
            }
        }
        flag
    }

    def setHomeArea(entity: IEntitySoulCustom,
                    x: Int,
                    y: Int,
                    z: Int,
                    maxDistance: Int) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.setHomeArea(entity, x, y, z, maxDistance)
        }
    }

    def getMaxHomeDistance(entity: IEntitySoulCustom): Float = {
        var flag = 0.0F
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val dist = trt.getMaxHomeDistance(entity)
            if (dist != 0.0F) {
                flag = dist
            }
        }
        flag
    }

    def detachHome(entity: IEntitySoulCustom) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.detachHome(entity)
        }
    }

    def hasHome(entity: IEntitySoulCustom): Boolean = {
        var flag = false
        for (trt <- SoulHelper.traitRegistry.getTraits if trt.hasHome(entity)) {
            flag = true
        }
        flag
    }

    def writeToNBT(entity: IEntitySoulCustom, compound: NBTTagCompound) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.writeToNBT(entity, compound)
        }
    }

    def readFromNBT(entity: IEntitySoulCustom, compound: NBTTagCompound) {
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            trt.readFromNBT(entity, compound)
        }
    }

    def getEntityTexture(entity: IEntitySoulCustom): String = {
        var flag: String = null
        for (trt <- SoulHelper.traitRegistry.getTraits) {
            val tex = trt.getEntityTexture(entity)
            if (tex != null && tex != "") {
                flag = tex
            }
        }
        flag
    }


    def onStruckByLightning(entity: IEntitySoulCustom, lightingBolt: EntityLightningBolt) {
        for(trt <- SoulHelper.traitRegistry.getTraits) {
            trt.onStruckByLightning(entity, lightingBolt)
        }
    }
}